functional-tests/fn_bridge_withdraw_happy.py

```diff
@@ -46,210 +42,292 @@ def __init__(self, ctx: flexitest.InitContext):
+def estimate_withdraw_gas(web3: Web3, el_address, change_address_pk, deposit_amount):
+    """
+    Estimate the gas for the withdrawal transaction.
+    """
+
+    data_bytes = bytes.fromhex(change_address_pk)
+
+    transaction = {
+        "from": el_address,
+        "to": PRECOMPILE_BRIDGEOUT_ADDRESS,
+        "value": deposit_amount * SATS_TO_WEI,
+        "data": data_bytes,
+    }
+    return web3.eth.estimate_gas(transaction)
+
+
+def mine_blocks_until_maturity(
+    btcrpc,
+    withdraw_address,
+    btc_url,
+    btc_user,
+    btc_password,
+    original_balance,
+    number_of_blocks=12,
+):
+    """
+    Mine blocks until the withdraw address has a positive balance
+    By default, the number of blocks to mine is 12:
+    - 6 blocks to mature the DRT
+    - 6 blocks to mature the DT
+    """
+    btcrpc.proxy.generatetoaddress(number_of_blocks, UNSPENDABLE_ADDRESS)
+    utils.wait_until(
+        lambda: get_balance(withdraw_address, btc_url, btc_user, btc_password) > original_balance,
+        timeout=10,
+    )
+
+
+def setup_services(ctx, debug_fn=print):
+    """
+    code to set up all necessary services and references:
+      - The Bitcoin service (btc) and the bitcoind RPC (btcrpc)
+      - The Sequencer service (seq) and the sequencer RPC (seqrpc)
+      - The Reth service (reth) and the reth RPC (rethrpc)
+      - The Web3 instance with signing middleware injected
+    """
+    btc = ctx.get_service("bitcoin")
+    seq = ctx.get_service("sequencer")
+    reth = ctx.get_service("reth")
+
+    seqrpc = seq.create_rpc()
+    btcrpc: BitcoindClient = btc.create_rpc()
+    rethrpc = reth.create_rpc()
+
+    # Debug info
+    if debug_fn:
+        debug_fn(f"BTC URL: {btcrpc.base_url}")
+        debug_fn(f"BTC user: {btc.props['rpc_user']}")
+        debug_fn(f"BTC password: {btc.props['rpc_password']}")
+        debug_fn(f"Sequencer Address: {seq.get_prop('address')}")
+
+    web3: Web3 = reth.create_web3()
+    eth_account = web3.eth.account.from_key(ETH_PRIVATE_KEY)
+    el_address = eth_account.address
+
+    # Inject signing middleware
+    web3.middleware_onion.inject(
+        SignAndSendRawMiddlewareBuilder.build(eth_account),
+        layer=0,
+    )
+
+    return btc, seq, reth, seqrpc, btcrpc, rethrpc, web3, el_address
+
+
+def check_initial_eth_balance(rethrpc, address, debug_fn=print):
+    """Asserts that the initial ETH balance for `address` is zero."""
+    balance = int(rethrpc.eth_getBalance(address), 16)
+    if debug_fn:
+        debug_fn(f"Strata Balance before deposits: {balance}")
+    assert balance == 0, "Strata balance is not expected (should be zero initially)"
+
+
+def deposit_twice(ctx, web3, seqrpc, rethrpc, el_address, debug_fn=print):
+    """
+    Make two DRT deposits to ensure the EL address has enough funds for gas
+    and for subsequent withdrawals. Wait until the deposit is reflected on L2.
+    """
+    # Get bridge pubkey
+    bridge_pk = utils.get_bridge_pubkey(seqrpc)
+    if debug_fn:
```

Comment:
Can we not just omit this if clause? Looks annoying. The default value in the `debug_fn` argument indicates that this clause will always return true, unless `None` is passed(but this can be avoided by enforcing argument type).

Reply:
In that case, I second the approach by @bewakes i.e., annotating with `Callable[str, None]` perhaps and removing the `if` clause.

Reply:
These functions with `debug_fn` are no longer valid and has been covered by `BridgeTestBase` which has access to self.debug itself